### PR TITLE
First pass at focusring

### PR DESCRIPTION
**Pull request by robdodson**
This is a PR against my own copy of the CSSWG spec. I've used language from the [focusring spec](https://github.com/WICG/modality/blob/master/spec/focusring.bs) and tried to make it fit in with the existing [:focus section](https://www.w3.org/TR/selectors4/#the-focus-pseudo). @alice @bkardell PTAL

### CODE REVIEW

**Comment by bkardell** [comment]
"in the absence of UA or author stylesheets"  is that true? I don't think it is, without UA stylesheets the whole web falls apart, right? If unfamilliar, see "thinking about how css works" under [in this article](http://radar.oreilly.com/2015/03/pandoras-box-model.html ).  If familliar, sorry, help me understand that?

**Comment by bkardell** [comment]
seems inconsistent how notes are in that spec, see https://github.com/robdodson/csswg-drafts/blob/babbd8208cc5c5d5424ccf6aafb019f8a3f2c144/selectors/Overview.bs#L2021 for another example - I'm not sure which is the right one to follow.  I'm wondering whether it is worth trying to sum up something in the original non-normative piece in https://github.com/WICG/modality/blob/master/spec/focusring.bs#L73 because it seems to me we would actually like some experimentation going on there.  Eventually we need to get a future version of this that allows users opting into heuristic behavior so that, for example, a custom input can be 'texty' automatically.

**Comment by robdodson** [comment]
Seems like it's ok to use `class=note` based on the conformance section https://www.w3.org/TR/selectors4/#conformance

**Comment by bkardell** [comment]
seems p good to me?

File: selectors/Overview.bs [hunk 1]
```diff
@@ -2031,6 +2036,49 @@ The Input Focus Pseudo-class: '':focus''</h3>
...
+	</div>
+
+	<em>This section is non-normative, and is meant to help clarify when a
+	:focusring selector might match.</em>
+
+	A user agent might choose to use the following heuristic for deciding when
+	to apply '':focusring'' to elements that are not native interaction elements:
+
+	When focussing an element:
+	*   if the last interaction was a keyboard event, and the interaction
```
Comment (alice) [inline selectors/Overview.bs]: "if the element received focus via a keyboard interaction, including indirectly such as triggering a dialog via pressing a button using the keyboard"

File: selectors/Overview.bs [hunk 2]
```diff
@@ -2031,6 +2036,49 @@ The Input Focus Pseudo-class: '':focus''</h3>
...
+		For example, most UAs choose to display focusrings on native text fields
+		whenever they are focused, and on native buttons only when selected via
+		the keyboard.
+	</div>
+
+	<em>This section is non-normative, and is meant to help clarify when a
+	:focusring selector might match.</em>
+
+	A user agent might choose to use the following heuristic for deciding when
+	to apply '':focusring'' to elements that are not native interaction elements:
```
Comment (alice) [inline selectors/Overview.bs]: "to match '':focusring'' on focused elements which do not have any native focus ring behaviour defined" ?

File: selectors/Overview.bs [hunk 3]
```diff
@@ -2031,6 +2036,49 @@ The Input Focus Pseudo-class: '':focus''</h3>
...
+
+	When focussing an element:
+	*   if the last interaction was a keyboard event, and the interaction
+			was less than 100ms in the past, apply '':focusring''.
+
+	When unfocussing an element:
```
Comment (alice) [inline selectors/Overview.bs]: Kinda goes without saying?

File: selectors/Overview.bs [hunk 4]
```diff
@@ -2031,6 +2036,49 @@ The Input Focus Pseudo-class: '':focus''</h3>
...
+	*   stop applying '':focusring''.
+
+	When handling a keyboard event:
+	*   if a non-native element is focussed, apply '':focusring'' to that element.
```
Comment (alice) [inline selectors/Overview.bs]: ...element, even if it did not become focused via the keyboard.

File: selectors/Overview.bs [hunk 5]
```diff
@@ -2031,6 +2036,49 @@ The Input Focus Pseudo-class: '':focus''</h3>
...
+
+	<div class='note'>
+		Authors who build non-native interactive elements that are text-field
```
Comment (alice) [inline selectors/Overview.bs]: This makes me a little sad but it makes sense.

File: selectors/Overview.bs (hunk 1 above)
Comment (robdodson) [inline selectors/Overview.bs]: should I leave in the 100ms bit?

Comment (alice) [inline selectors/Overview.bs]: nope it's an implementation detail IMO

**Comment by bkardell** [comment]
read this sentence carefully https://github.com/robdodson/csswg-drafts/pull/1/files#diff-b8f9f5846b8094cd2099649b6fb92111R2073

does it make sense? I think there is an is that doesn't belong there, or something got tweaked?

File: selectors/Overview.bs (hunk 5 above)
Comment (bkardell) [inline selectors/Overview.bs]: I really want this to not be a thing... this is gonna get such reactions, I really wish we would just address it now :(

File: selectors/Overview.bs [hunk 6]
```diff
@@ -2031,6 +2036,49 @@ The Input Focus Pseudo-class: '':focus''</h3>
...
+	ring behaviour defined:
+
+	When focusing an element:
+	*   if the element received focus via a keyboard interaction, including
+	indirectly such as triggering a dialog via pressing a button using the
+	keyboard, apply '':focusring''.
+
+	When handling a keyboard event:
+	*   if a non-native element, even if it did not become focused via the
+	keyboard, is focussed, apply '':focusring'' to that element.
```
Comment (alice) [inline selectors/Overview.bs]: lol, I should have thought harder about my suggestion.
'''
if a non-native element is focused, even if it did not become focused via the keyboard, apply '':focusring'' to that element.
'''

Comment (bkardell) [inline selectors/Overview.bs]: Would it just be simpler to say 
"""
No matter how they are focused, always apply ":focusring" to non-native elements with the focus.
"""

**Pull request merged**